modules/phase_field/src/actions/GrandPotentialKernelAction.C

```diff
@@ -244,50 +261,133 @@ GrandPotentialKernelAction::act()
     _problem->addKernel("ACGrGrMulti", kernel_name, params);
   } // for (unsigned int i = 0; i < n_etas + n_grs; ++i)
 
-  // Chemical Potentials
-  for (unsigned int i = 0; i < n_w; ++i)
+  if (mass_conservation) // mass conservation kernels with conc and chempot coupling
   {
-    // SusceptibilityTimeDerivative
-    InputParameters params = _factory.getValidParams("SusceptibilityTimeDerivative");
-    params.set<NonlinearVariableName>("variable") = w_names[i];
-    params.set<MaterialPropertyName>("f_name") = chis[i];
-    params.set<std::vector<VariableName>>("coupled_variables") = v0;
-    params.set<bool>("implicit") = implicity;
-    params.set<bool>("use_displaced_mesh") = displaced_mesh;
-    kernel_name = "ChiDt_" + w_names[i];
-    _problem->addKernel("SusceptibilityTimeDerivative", kernel_name, params);
+    std::vector<VariableName> v3;
+    v3.resize(n_w);
+    for (unsigned int i = 0; i < n_w; ++i)
+      v3[i] = w_names[i];
+    c_names = getParam<std::vector<NonlinearVariableName>>("concentrations");
+    unsigned int n_c = c_names.size();
+    std::vector<VariableName> v4;
+    v4.resize(n_c);
+    for (unsigned int i = 0; i < n_c; ++i)
+      v4[i] = c_names[i];
+    for (unsigned int i = 0; i < n_c; ++i)
+    {
+      // TimeDerivative concentration Kernel
+      InputParameters params = _factory.getValidParams("TimeDerivative");
+      params.set<NonlinearVariableName>("variable") = c_names[i];
+      params.set<bool>("implicit") = implicity;
+      params.set<bool>("use_displaced_mesh") = displaced_mesh;
+      kernel_name = "DT_" + c_names[i];
+      _problem->addKernel("TimeDerivative", kernel_name, params);
 
-    // MatDiffusion
-    params = _factory.getValidParams("MatDiffusion");
-    params.set<NonlinearVariableName>("variable") = w_names[i];
-    params.set<bool>("implicit") = implicity;
-    params.set<bool>("use_displaced_mesh") = displaced_mesh;
-    params.set<MaterialPropertyName>("diffusivity") = M[i];
-    kernel_name = "MatDif_" + w_names[i];
-    if (aniso.get(i))
-      _problem->addKernel("MatAnisoDiffusion", kernel_name, params);
-    else
+      // MatDiffusion concentration (coupled with chempot)
+      params = _factory.getValidParams("MatDiffusion");
+      params.set<NonlinearVariableName>("variable") = c_names[i];
+      params.set<std::vector<VariableName>>("v") = v3;
```

Comment:
```suggestion
      params.set<std::vector<VariableName>>("v") = w_names;
```

same for `v4`. This is just a suggestion. It seems to me that those vectors are not necessary. FYI, you could also just copy a vactor by assignment: `auto v3 = w_names;` . No need for the loops.